src/corecel/cont/Bitset.hh

```diff
@@ -0,0 +1,553 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/cont/Bitset.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <climits>
+
+#include "corecel/Config.hh"
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/cont/detail/BitsetUtils.hh"
+#include "corecel/math/Algorithms.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Device-compatible bitset implementation.
+ *
+ * This implementation is based on libstdc++'s std::bitset implementation.
+ * It it a subset of the C++ standard, but it should be sufficient
+ * for our current use case. Given that GPU typically use 32-bit words, this
+ * uses unsigned int as the word type instead of the unsigned long used by the
+ * standard library. This container is not thread-safe, multiple threads are
+ * likely to manipulate the same word and access is not synchronized.
+ *
+ * The following methods are not implemented:
+ * - conversions to string, to_ulong, to_ullong
+ * - stream operators
+ * - shift operators
+ * - hash support
+ * - construct from string, from_ullong
+ */
+template<size_type N>
+class Bitset
+{
+    static_assert(N > 0, "Zero-sized bitset is not supported");
+
+  public:
+    //!@{
+    //! \name Type aliases
+    using word_type = unsigned int;
+    //!@}
+
+    class reference;
+
+    static constexpr size_type bits_per_word = CHAR_BIT * sizeof(word_type);
+    static constexpr size_type num_words = ceil_div(N, bits_per_word);
+
```

Comment:
These don't need to be public, so perhaps they'd be better as private static data with a `_` suffix?